server/push/fcm/push_fcm.go

```diff
@@ -101,42 +113,113 @@ func (Handler) Init(jsonconf string) error {
 }
 
 func sendNotifications(rcpt *push.Receipt, config *configType) {
-	ctx := context.Background()
 	messages := PrepareNotifications(rcpt, &config.Android)
-	if len(messages) == 0 {
+	n := len(messages)
+	if n == 0 {
 		return
 	}
 
-	for _, m := range messages {
-		_, err := handler.client.Send(ctx, m.Message)
+	ctx := context.Background()
+	for i := 0; i < n; i += pushBatchSize {
+		upper := i + pushBatchSize
+		if upper > n {
+			upper = n
+		}
+		var batch []*fcm.Message
+		for j := i; j < upper; j++ {
+			batch = append(batch, messages[j].Message)
+		}
+		resp, err := handler.client.SendAll(ctx, batch)
 		if err != nil {
-			if fcm.IsMessageRateExceeded(err) ||
-				fcm.IsServerUnavailable(err) ||
-				fcm.IsInternal(err) ||
-				fcm.IsUnknown(err) {
-				// Transient errors. Stop sending this batch.
-				log.Println("fcm transient failure", err)
-				return
-			}
+			// Complete failure.
+			log.Println("fcm SendAll failed", err)
+			break
+		}
 
-			if fcm.IsMismatchedCredential(err) || fcm.IsInvalidArgument(err) {
-				// Config errors
-				log.Println("fcm push: failed", err)
-				return
-			}
+		// Check for partial failure.
+		if handlePushErrors(resp, messages[i:upper]) {
+			break
+		}
+	}
+}
 
-			if fcm.IsRegistrationTokenNotRegistered(err) {
-				// Token is no longer valid.
-				log.Println("fcm push: invalid token", err)
-				err = store.Devices.Delete(m.Uid, m.DeviceId)
-				if err != nil {
-					log.Println("fcm push: failed to delete invalid token", err)
-				}
-			} else {
-				log.Println("fcm push:", err)
-			}
+func processSubscription(req *push.ChannelReq) {
+	if len(req.Devices) == 0 {
+		return
+	}
+
+	if len(req.Devices) > subBatchSize {
+		req.Devices = req.Devices[0:subBatchSize]
+	}
+
+	var err error
+	var resp *fcm.TopicManagementResponse
+	if req.Unsub {
+		resp, err = handler.client.UnsubscribeFromTopic(context.Background(), req.Devices, req.Channel)
+	} else {
+		resp, err = handler.client.SubscribeToTopic(context.Background(), req.Devices, req.Channel)
+	}
+	if err != nil {
+		// Complete failure.
+		log.Println("fcm sub or upsub failed", req.Unsub, err)
+	} else {
+		// Check for partial failure.
+		handleSubErrors(resp, req.Uid, req.Devices)
+	}
+}
+
+// handlePushError processes errors returned by a call to fcm.SendAll.
+// returns true to stop further processing of other messages.
```

Comment:
Perhaps, it'll be somewhat more readable to flip the return value semantics:
I read handlePushErrors() return whether errors were handled correctly and we can proceed with more pushes.

Reply:
Done